presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarJoinUtil.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanVisitor;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+
+public final class ScalarJoinUtil
+{
+    private ScalarJoinUtil() {}
+
+    public static boolean isScalarJoin(JoinNode node)
+    {
+        return (node.getType() == JoinNode.Type.INNER) &&
+                node.getRight().accept(new IsScalarJoinPlanVisitor(), null);
+    }
+
+    private static final class IsScalarJoinPlanVisitor
+            extends PlanVisitor<Void, Boolean>
+    {
+        @Override
+        protected Boolean visitPlan(PlanNode node, Void context)
+        {
+            return false;
+        }
+
+        @Override
+        public Boolean visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            return true;
+        }
+
+        @Override
+        public Boolean visitExchange(ExchangeNode node, Void context)
+        {
+            return (node.getType() == ExchangeNode.Type.REPLICATE) &&
```

Comment:
why do you need to check Exchange nodes? What if plan contains nested Exchanges where the nested one is the one related to subquery? In that case it might  return `false`, is that what you want?